File: tslib-py/tslib/diagnostics.py
from typing import List

from ._http import _Session
from .models import LjungBoxResult


class DiagnosticsAPI:
    def __init__(self, session: _Session) -> None:
        self._s = session

    def ljung_box(
        self,
        residuals: List[float],
        *,
        lags: int = 10,
        degrees_of_freedom_adjustment: int = 0,
    ) -> LjungBoxResult:
        body = {
            "residuals": residuals,
            "lags": lags,
            "degreesOfFreedomAdjustment": degrees_of_freedom_adjustment,
        }
        return LjungBoxResult._from(self._s.post("/api/diagnostics/ljung-box", body))

File: tslibR/R/diagnostics.R
#' Ljung-Box portmanteau test
#'
#' Tests whether a sequence of residuals shows significant autocorrelation up
#' to the given number of lags. A small p-value (< 0.05) suggests the residuals
#' are not white noise.
#'
#' @param client A [`tslib_client`] object.
#' @param residuals Numeric vector of model residuals.
#' @param lags Number of lags to include in the test statistic (default 10).
#' @param df_adjustment Degrees-of-freedom adjustment for estimated model
#'   parameters (e.g. p + q for an ARIMA model) (default 0).
#' @return Named list with `statistic`, `pValue`, `lags`, and
#'   `rejectsAtFivePercent` (logical).
#' @export
tslib_ljung_box <- function(client, residuals,
                            lags          = 10L,
                            df_adjustment = 0L) {
  body <- list(
    residuals                    = as.list(residuals),
    lags                         = lags,
    degreesOfFreedomAdjustment   = df_adjustment
  )
  .tslib_post(client, "/api/diagnostics/ljung-box", body)
}
